nixos/home-assistant: harden systemd service

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

```
This is what is still exposed, and it should still allow things to work
as usual.

✗ PrivateNetwork=                                             Service has access to the host's network                                            0.5
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                               0.3
✗ CapabilityBoundingSet=~CAP_NET_ADMIN                        Service has network configuration privileges                                        0.2
✗ DeviceAllow=                                                Service has no device ACL                                                           0.2
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                    0.2
✗ PrivateDevices=                                             Service potentially has access to hardware devices                                  0.2
✗ PrivateUsers=                                               Service has access to other users                                                   0.2
✗ AmbientCapabilities=                                        Service process receives ambient capabilities                                       0.1
✗ CapabilityBoundingSet=~CAP_NET_(BIND_SERVICE|BROADCAST|RAW) Service has elevated networking privileges                                          0.1
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                       0.1
✗ RestrictAddressFamilies=~AF_UNIX                            Service may allocate local sockets                                                  0.1

→ Overall exposure level for home-assistant.service: 1.9 OK 🙂
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mweinelt committed:
nixos/home-assistant: harden systemd service

This is what is still exposed, and it should still allow things to work
as usual.

✗ PrivateNetwork=                    Service has access to the host's …      0.5
✗ RestrictAddressFamilies=~AF_(INET… Service may allocate Internet soc…      0.3
✗ DeviceAllow=                       Service has a device ACL with som…      0.1
✗ IPAddressDeny=                     Service does not define an IP add…      0.2
✗ PrivateDevices=                    Service potentially has access to…      0.2
✗ PrivateUsers=                      Service has access to other users       0.2
✗ SystemCallFilter=~@resources       System call allow list defined fo…      0.2
✗ RootDirectory=/RootImage=          Service runs within the host's ro…      0.1
✗ SupplementaryGroups=               Service runs with supplementary g…      0.1
✗ RestrictAddressFamilies=~AF_UNIX   Service may allocate local sockets      0.1

→ Overall exposure level for home-assistant.service: 1.6 OK :-)

This can grow to as much as ~1.9 if you use one of the bluetooth or nmap
trackers or the emulated_hue component, all of which required elevated
permisssions.

mweinelt commented:
```
14:27 <hexa-> is setting +eip (in any combination) on a capability in a systemd unit useful?
14:27 <hexa-> like from what I understand CapabilityBoundingSet is basically the "p" (permitted)
14:27 <hexa-> and AmbientCapabilities could be interpreted as "e" (effective)
14:29 <hexa-> the manpage only states "Note that this option also affects the respective capabilities in the effective, permitted and inheritable capability sets."
14:30 <hexa-> so I feel like there is no point in setting +eip or any combination thereof
```

I've removed `+eip` from `cap_net_admin` for this reason. Sadly I believe @peterhoeg, who originally added those capabilities (presumably due to the bluetooth tracker component) migrated to OpenHAB. So we'd need someone else to test or confirm this.

Mic92 commented on nixos/modules/services/misc/home-assistant.nix:
```diff
@@ -251,16 +251,54 @@ in {
+        KillSignal = "SIGINT";
+
+        # Hardening
+        AmbientCapabilities = lib.mkForce [
```
Is it feasible to only add these if said module is actually enabled?

mweinelt replied:
Yeah, could be. The `lib.mkForce` is a Copy/paste remainder of my local test setup.

mweinelt replied:
Done in [`4a21fd3` (#120526)](https://github.com/NixOS/nixpkgs/pull/120526/commits/4a21fd352959cc5aca413869f671b46366e22d13).

Mic92 commented on nixos/modules/services/misc/home-assistant.nix:
```diff
@@ -251,16 +251,54 @@ in {
+          "cap_net_admin"
+        ];
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
```
Do we have any node.js plugins around? This would crash them.

mweinelt replied:
I don't think we have.

mweinelt committed:
nixos/tests/home-assistant: test capability passing

Configures the emulated_hue component and expects CAP_NET_BIND_SERVICE
to be passed in order to be able to bind to 80/tcp.

Also print the systemd security analysis, so we can spot changes more
quickly.

mweinelt commented:
Capabilities are now conditional based on components used.

```
✗ PrivateNetwork=                                             Service has access to the host's network                                            0.5
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                               0.3
✗ DeviceAllow=                                                Service has no device ACL                                                           0.2
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                    0.2
✗ PrivateDevices=                                             Service potentially has access to hardware devices                                  0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                       0.1
✗ RestrictAddressFamilies=~AF_UNIX                            Service may allocate local sockets                                                  0.1

→ Overall exposure level for home-assistant.service: 1.3 OK 🙂
```

mweinelt commented:
Core dump at my parents setup.

```
(gdb) bt full
#0  0x00007fdc8c223299 in syscall () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6
No symbol table info available.
#1  0x00007fdc52616cd6 in alloc_mmap () from /nix/store/79491jwy14lbfrhbwwl6jr6qnkkrrwb8-lapack-3/lib/liblapack.so.3
No symbol table info available.
#2  0x00007fdc526170d2 in blas_memory_alloc () from /nix/store/79491jwy14lbfrhbwwl6jr6qnkkrrwb8-lapack-3/lib/liblapack.so.3
No symbol table info available.
#3  0x00007fdc52617765 in adjust_thread_buffers () from /nix/store/79491jwy14lbfrhbwwl6jr6qnkkrrwb8-lapack-3/lib/liblapack.so.3
No symbol table info available.
#4  0x00007fdc52617bf0 in blas_thread_init () from /nix/store/79491jwy14lbfrhbwwl6jr6qnkkrrwb8-lapack-3/lib/liblapack.so.3
No symbol table info available.
#5  0x00007fdc523e507f in gotoblas_init () from /nix/store/79491jwy14lbfrhbwwl6jr6qnkkrrwb8-lapack-3/lib/liblapack.so.3
No symbol table info available.
#6  0x00007fdc8c8be9ee in call_init () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/ld-linux-x86-64.so.2
No symbol table info available.
#7  0x00007fdc8c8bead4 in _dl_init () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/ld-linux-x86-64.so.2
No symbol table info available.
#8  0x00007fdc8c25f0bc in _dl_catch_exception () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6
No symbol table info available.
#9  0x00007fdc8c8c2d88 in dl_open_worker () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/ld-linux-x86-64.so.2
No symbol table info available.
#10 0x00007fdc8c25f075 in _dl_catch_exception () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6
No symbol table info available.
#11 0x00007fdc8c8c25dd in _dl_open () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/ld-linux-x86-64.so.2
No symbol table info available.
#12 0x00007fdc8c500246 in dlopen_doit () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libdl.so.2
No symbol table info available.
#13 0x00007fdc8c25f075 in _dl_catch_exception () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6
No symbol table info available.
#14 0x00007fdc8c25f10f in _dl_catch_error () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6
No symbol table info available.
#15 0x00007fdc8c5008f5 in _dlerror_run () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libdl.so.2
No symbol table info available.
#16 0x00007fdc8c5002c6 in dlopen@@GLIBC_2.2.5 () from /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libdl.so.2
No symbol table info available.
#17 0x00007fdc8c700533 in _PyImport_FindSharedFuncptr () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#18 0x00007fdc8c6ca4ac in _PyImport_LoadDynamicModuleWithSpec () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#19 0x00007fdc8c6c7f95 in _imp_create_dynamic () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#20 0x00007fdc8c609323 in cfunction_vectorcall_FASTCALL () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#21 0x00007fdc8c5c518f in PyVectorcall_Call () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#22 0x00007fdc8c59dd68 in _PyEval_EvalFrameDefault () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#23 0x00007fdc8c6a30af in _PyEval_EvalCodeWithName () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#24 0x00007fdc8c5c30d1 in _PyFunction_Vectorcall () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
No symbol table info available.
#25 0x00007fdc8c595dad in call_function () from /nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/libpython3.8.so.1.0
```

mweinelt commented:
Everything that wants numpy relies on lapack/blas and that wants mmap, therefore I allowed the `@resources` group, which contains the mmap syscall.

```
✗ PrivateNetwork=                                             Service has access to the host's network                                                                 0.5
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                                                    0.3
✗ DeviceAllow=                                                Service has no device ACL                                                                                0.2
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                                         0.2
✗ PrivateDevices=                                             Service potentially has access to hardware devices                                                       0.2
✗ SystemCallFilter=~@resources                                System call allow list defined for service, and @resources is included (e.g. ioprio_set is allowed)      0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                                            0.1
✗ MemoryDenyWriteExecute=                                     Service may create writable executable memory mappings                                                   0.1
✗ RestrictAddressFamilies=~AF_UNIX                            Service may allocate local sockets                                                                       0.1

→ Overall exposure level for home-assistant.service: 1.5 OK 🙂
```

Mic92 commented:
Are you sure `mmap` was the actual issue there? You cannot even run the link-loader without mmap.

mweinelt commented:
I guessed as much from the call trace up here https://github.com/NixOS/nixpkgs/pull/120526#issuecomment-826359009.

mweinelt commented:
> I also gave PrivateUsers a shot, though I could imagine that to be breaking things, like file permissions here and there.

Yeah, so this broke privilege elevation through capabilities. Let's leave this out for now.

mweinelt committed:
nixos/tests/home-assistant: update maintainership to home-assistant team

mweinelt committed:
nixos/home-assistant: Restart systemd unit on restart service

Home-assistant through its `--runner` commandline flag supports sending
exit code 100 when the `homeassistant.restart` service is called.

With `RestartForceExitStatus` we can listen for that specific exit code
and restart the whole systemd unit, providing an actual clean restart
with fresh processes. Additional treat exit code 100 as a successful
termination.

mweinelt commented:
@GrahamcOfBorg test home-assistant

Mic92 commented:
> > I also gave PrivateUsers a shot, though I could imagine that to be breaking things, like file permissions here and there.
> 
> Yeah, so this broke privilege elevation through capabilities. Let's leave this out for now.

I also don't really consider this a security feature to be honest.

mweinelt commented:
@Mic92 @fabaff @globin @etu @mvnetbiz

Can you guys test-drive this change, or do you trust me enough not to break your setup? :smiling_imp:

mvnetbiz commented:
I'll check it out now.

mweinelt commented on nixos/modules/services/misc/home-assistant.nix:
```diff
@@ -245,22 +245,71 @@ in {
+          "CAP_NET_ADMIN"
+          "CAP_NET_BIND_SERVICE"
+          "CAP_NET_RAW"
+        ]));
```
FWIW: I came up with this list by grepping the home-assistant website.

We've covered the bluetooth and nmap trackers as well as emulated_hue.

We don't have support for `ps4`, `eddystone_temperature` and `kulersky`

I'm not quite sure how we can handle the capabilities on `hcitool`, is anyone using bluetooth trackers?

```
❯ rg "\\bcap_"
source/_integrations/bluetooth_le_tracker.markdown
80:sudo setcap 'cap_net_raw,cap_net_admin+eip' $(readlink -f $(which python3))
81:sudo setcap 'cap_net_raw+ep' $(readlink -f $(which hcitool))

source/_integrations/dwd_weather_warnings.markdown
36:A list of valid warncell ids and names can be found at https://www.dwd.de/DE/leistungen/opendata/help/warnungen/cap_warncellids_csv.html.

source/_integrations/emulated_hue.markdown
170:sudo setcap 'cap_net_bind_service=+ep' /srv/homeassistant/homeassistant_venv/bin/python3

source/_integrations/eddystone_temperature.markdown
28:sudo setcap 'cap_net_raw,cap_net_admin+eip' $(readlink -f $(which python3))

source/_integrations/nmap_tracker.markdown
95:sudo setcap cap_net_raw,cap_net_admin,cap_net_bind_service+eip /usr/bin/nmap

source/_integrations/kulersky.markdown
38:sudo setcap 'cap_net_raw,cap_net_admin+eip' `readlink -f \`which python3\``
39:sudo setcap 'cap_net_raw+ep' `readlink -f \`which hcitool\``

source/_integrations/ps4.markdown
67:sudo setcap 'cap_net_bind_service=+ep' <python>
75:sudo setcap 'cap_net_bind_service=+ep' /usr/bin/python3.5
```

mweinelt replied:
I gave the `nmap_tracker` a shot and it looks to be working like this.

mvnetbiz commented:
Any idea what caps are needed for ZHA? (/dev/ttyUSBx access)

mweinelt commented:
> Any idea what caps are needed for ZHA? (/dev/ttyUSBx access)

My first thought is `SupplementoryGroups = [ "dialout" ];`

mvnetbiz commented:
Figured it out, I had to add
```
{
  systemd.services.home-assistant.serviceConfig.DeviceAllow = [
    "/dev/serial/by-id/usb-Silicon_Labs_HubZ_Smart_Home_Controller_3120005E-if01-port0 rw"
    "/dev/serial/by-id/usb-Silicon_Labs_HubZ_Smart_Home_Controller_3120005E-if00-port0 rw"
  ];
}
```
DeviceAllow doesn't actually show up in the unit, but it is actually implied by `ProtectClock=true`, which I guess does `DeviceAllow=char-rtc r` behind the scenes. That was a bit confusing. It doesn't show up in `systemd analyze security` either, but maybe the purpose of that is to show what is explicitly done? I wonder if there is a way to show the actual enforced set of rules.

mvnetbiz reviewed (approved):
Works for me, but is there a way to warn that new settings might be needed for certain configurations? Should something like DeviceAllow be added to an explicit option in `services.home-assistant`?

mweinelt commented:
> Works for me, but is there a way to warn that new settings might be needed for certain configurations? Should something like DeviceAllow be added to an explicit option in `services.home-assistant`?

That's somewhat disappointing. I was previously of the opinion, that I don't want to get into the rats nest of setting `DeviceAllow=`, but now I am reevaluating that. We could set to something along the lines of `DeviceAllow=/dev/ttyUSB*,/dev/ttyACM*`. Another option would be to drop `ProtectClock=` in favor of `SystemCallFilter=~@clock`, so we wouldn't set `DeviceAllow=` after all.

> I wonder if there is a way to show the actual enforced set of rules.

Use `systemctl show home-assistant.service`.

mvnetbiz commented:
I'm not personally annoyed at having to set it but it's probably best to remove it. I'm not sure what other devices I might ever connect to the same machine, but either way I will just keep these options in my personal config now, so whether or not you remove it I will have it. I think it's probably not a huge security issue in most setups, and where it is an issue it can be manually set in order to not cause anyone problems.

mweinelt commented:
I have a change locally that should improve the situation quite a bit. I'm swamped right now, will push it later, needs some more finishing touches.

mvnetbiz commented:
Maybe we can come up with a whole list of integrations that use local devices, and then if any of these is used, warn if `DeviceAllow` is empty. It doesn't look like the NixOS module supports H-A downloading modules from pip, so anybody using integrations is either putting the config in nix, or overriding `extraPackages` so I think users of this module will expect to check their configuration if there is a problem. I'm not sure if there is an approachable way to come up with such a list though.

mvnetbiz commented:
Thanks, no hurry.

mweinelt commented:
> Nevermind, there's some components that may or may not use a local device and there's no way to tell which way ahead of time if it uses the GUI config flow.

Yeah, you always have to specify the component in your YAML config for its dependencies to be included. And I'm reusing the same logic for capabilities and devices now.

mvnetbiz commented:
I see there's possibilities like using the zha component but with a zigbee coordinator that provides an IP-zigbee bridge instead of local USB /serial, so in that case setting DeviceAllow would be unnecessary

mweinelt commented:
Hah, there is a limit to the complexity I'm willing to integrate here! :laughing:

mweinelt commented:
Okay, here we are. I've allowed certain character devices, that are used for UART and other protocols. At the same time I've migrated the `dialout` group from the user into the service as `SupplementaryGroups=`.

@mvnetbiz That should allow `zha` to work without setting explicit `DeviceAllow=` per device.

mweinelt commented:
And another change I had on that branch. Might as well add it here, it's minor.

Home-assistant supports signaling its supervisor, that it wants to restart, it sends exit code 100 for that. I added support for that using `RestartForceExitStatus=100` and made the termination using that exit code a success using `SuccessExitStatus=100`. This gives us a clean restart of the python interpreter.

Calling the `homeassistant.restart` service from within home-assistant will now restart the systemd unit.

```
hass[17946]: 2021-04-28 19:16:31 INFO (MainThread) [aioesphomeapi.connection] 192.168.120.4: Error while reading incoming messages: Error while receiving data: 0 bytes read on a total of 1 expected bytes
Apr 28 19:16:32 dione systemd[1]: home-assistant.service: Succeeded.
systemd[1]: home-assistant.service: Consumed 16.987s CPU time, received 266.4K IP traffic, sent 532.0K IP traffic.
systemd[1]: home-assistant.service: Scheduled restart job, restart counter is at 1.
dione systemd[1]: Stopped Home Assistant.
systemd[1]: home-assistant.service: Consumed 16.987s CPU time, received 266.4K IP traffic, sent 532.0K IP traffic.
systemd[1]: Starting Home Assistant...
systemd[1]: Started Home Assistant.
hass[18011]: 2021-04-28 19:16:36 WARNING (MainThread) [homeassistant.bootstrap] Skipping pip installation of required modules. This may cause issues

mvnetbiz commented:
New version works without manually adding `DeviceAllow=`. I tested Bluetooth tracker with a USB adapter and got some recognized devices, looks like this is needed.
```
        RestrictAddressFamilies = [
          "AF_UNIX"
          "AF_INET"
          "AF_INET6"
        ] ++ optionals (useComponent "bluetooth_tracker" || useComponent "bluetooth_le_tracker") [
          "AF_BLUETOOTH"
        ];
```
I don't have a Bluetooth LE adapter, so I can't test that.

mweinelt commented:
Indeed, that makes total sense. Pybluez opens an AF_BLUETOOTH socket from within Python. The Bluetooth LE tracker shelves out to hcitool, which likely does the same.

mvnetbiz reviewed (approved):
Tested with latest push, looks good to me.

mweinelt commented on nixos/modules/services/misc/home-assistant.nix:
```diff
@@ -245,22 +245,84 @@ in {
+        # List of capabilities to equip home-assistant with, depending on configured components
+        capabilities = [
+          # Empty string first, so we will never accidentally have an empty capability bounding set
+          ""
```
Went with an empty string as recommended in https://github.com/NixOS/nixpkgs/issues/120617#issuecomment-830685115.
